subsys/bluetooth/audio/cap_commander.c

```diff
@@ -399,14 +400,302 @@ int bt_cap_commander_broadcast_reception_start(
+}
+
+static void cap_commander_broadcast_assistant_recv_state_cb(
+	struct bt_conn *conn, int err, const struct bt_bap_scan_delegator_recv_state *state)
+{
+	struct bt_cap_common_proc *active_proc = bt_cap_common_get_active_proc();
+	struct bt_conn *cap_conn = NULL;
+	char le_addr[BT_ADDR_LE_STR_LEN];
+	uint16_t total_bis_sync;
+	uint8_t src_id = 0;
+
+	if (err != 0) {
+		LOG_DBG("BASS recv state read failed (%d)\n", err);
+		return;
+	}
+
+	if (state == NULL) {
+		/* Empty receive state */
+		return;
+	}
+
+	bt_addr_le_to_str(&state->addr, le_addr, sizeof(le_addr));
+
+	LOG_DBG("BASS recv state: src_id %u, addr %s, sid %u, sync_state %u", state->src_id,
+		le_addr, state->adv_sid, state->pa_sync_state);
+
+	if (state->encrypt_state == BT_BAP_BIG_ENC_STATE_BAD_CODE) {
+		LOG_DBG("Encryption state is BT_BAP_BIG_ENC_STATE_BAD_CODE");
+		return;
+	}
+
+	if (bt_cap_common_conn_in_active_proc(conn) &&
+	    active_proc->proc_type == BT_CAP_COMMON_PROC_TYPE_BROADCAST_RECEPTION_STOP) {
+		total_bis_sync = 0;
+		for (uint8_t i = 0; i < state->num_subgroups; i++) {
+			const struct bt_bap_bass_subgroup *subgroup = &state->subgroups[i];
+
+			total_bis_sync |= subgroup->bis_sync;
+		}
+		for (size_t i = 0; i < active_proc->proc_cnt; i++) {
+			if (state->src_id == active_proc->proc_param.commander[i]
+						     .broadcast_reception_stop.src_id &&
+			    conn == active_proc->proc_param.commander[i].conn) {
+				src_id = state->src_id;
+				cap_conn = conn;
+			}
+		}
+
+		if (total_bis_sync == 0 && cap_conn == conn) {
+			LOG_DBG("Removing source for conn %p", (void *)cap_conn);
+			err = bt_bap_broadcast_assistant_rem_src(cap_conn, src_id);
+			if (err != 0) {
+				LOG_DBG("Failed to perform broadcast reception stop for conn %p: "
+					"%d",
+					(void *)cap_conn, err);
+				bt_cap_common_abort_proc(cap_conn, err);
+				cap_commander_proc_complete();
+			}
+		}
+	}
+}
+
+static void cap_commander_broadcast_assistant_rem_src_cb(struct bt_conn *conn, int err)
+{
+	struct bt_cap_common_proc *active_proc = bt_cap_common_get_active_proc();
+	struct bt_bap_broadcast_assistant_mod_src_param mod_src_param = {0};
+
+	if (!bt_cap_common_conn_in_active_proc(conn)) {
+		/* State change happened outside of a procedure; ignore */
+		return;
+	}
+
+	if (err != 0) {
+		LOG_DBG("Failed removing source: %d", err);
+		LOG_DBG("Aborting the proc %d %d", active_proc->proc_done_cnt,
+			active_proc->proc_initiated_cnt);
+
+		bt_cap_common_abort_proc(conn, err);
+	} else {
+		active_proc->proc_done_cnt++;
+
+		LOG_DBG("Conn %p broadcast source removed (%zu/%zu streams done)", (void *)conn,
+			active_proc->proc_done_cnt, active_proc->proc_cnt);
+	}
+
+	if (bt_cap_common_proc_is_aborted()) {
+		if (bt_cap_common_proc_all_handled()) {
+			cap_commander_proc_complete();
+		}
+
+		return;
+	}
+
+	if (!bt_cap_common_proc_is_done()) {
+		struct bt_cap_commander_proc_param *proc_param;
+
+		proc_param = &active_proc->proc_param.commander[active_proc->proc_done_cnt];
```

Comment:
Is this resilient to server-induced remove source? 

For example if we are removing the source for scan delegators SD_1, SD_2 and SD_3, and we just finished removing for SD_1 and started removing for SD_2 but SD_3 removed the source by itself, then then we increment `proc_done_cnt` here then there is no remove state to remove. 

We had a similar issue for the CAP Initiator until we implemented the `update_proc_done_cnt` function, which does state checks for each procedure rather than a simple counter, which allows it to better do the right thing if the remote devices are modified during the procedure

Reply:
ping @kruithofa

Reply:
as discussed, issue #78166 has been created